v2/internal/controller/controllers/targetnamespace_test.go

```diff
@@ -0,0 +1,258 @@
+
+const (
+	timeoutFast = time.Minute * 3
+	retry       = time.Second * 3
+)
+
+func TestTargetNamespaces(t *testing.T) {
+	t.Parallel()
+	tc := globalTestContext.ForTest(t)
+
+	createNamespaces(tc, "watched", "unwatched")
+	configuredNamespaces := os.Getenv("AZURE_TARGET_NAMESPACES")
+	podNamespace := os.Getenv("POD_NAMESPACE")
+
+	// We can't check for operator namespace if there's not one set.
+	tc.Expect(podNamespace).ToNot(Equal(""), "no POD_NAMESPACE set")
+
+	standardSpec := resources.ResourceGroupSpec{
+		Location: tc.AzureRegion,
+		Tags:     testcommon.CreateTestResourceGroupDefaultTags(),
+	}
+	// Create resource groups in these two namespaces - we can't
+	// easily use the test context for this because that's geared to
+	// creating things in one test namespace, while we need fixed
+	// namespaces.
+	rgDefault := resources.ResourceGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      tc.Namer.GenerateName("rg"),
+			Namespace: "default",
+		},
+		Spec: standardSpec,
+	}
+	tc.CreateResourceGroupAndWait(&rgDefault)
+	// Check that the instance is annotated with the operator namespace.
+	checkNamespaceAnnotation(tc, &rgDefault, podNamespace)
+
+	// The watched namespace is also reconciled.
+	rgWatched := resources.ResourceGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      tc.Namer.GenerateName("rg"),
+			Namespace: "watched",
+		},
+		Spec: standardSpec,
+	}
+	tc.CreateResourceGroupAndWait(&rgWatched)
+	checkNamespaceAnnotation(tc, &rgWatched, podNamespace)
+
+	// But the unwatched namespace isn't...
+	unwatchedName := tc.Namer.GenerateName("rg")
+	rgUnwatched := resources.ResourceGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      unwatchedName,
+			Namespace: "unwatched",
+		},
+		Spec: standardSpec,
+	}
+	_, err := tc.CreateResourceGroup(&rgUnwatched)
+	tc.Expect(err).ToNot(HaveOccurred())
+
+	name := types.NamespacedName{Name: unwatchedName, Namespace: "unwatched"}
+
+	gotNamespaceAnnotation := func(g Gomega) bool {
+		var instance resources.ResourceGroup
+		ctx := context.Background()
+		err := tc.KubeClient.Get(ctx, name, &instance)
+		g.Expect(err).NotTo(HaveOccurred())
+		res, err := meta.Accessor(&instance)
+		g.Expect(err).NotTo(HaveOccurred())
+		return res.GetAnnotations()[controllers.NamespaceAnnotation] == podNamespace
+	}
+
+	if configuredNamespaces == "" {
+		t.Log("**** all namespaces mode")
+		// The operator should be watching all namespaces.
+		tc.G.Eventually(
+			gotFinalizer(tc, name),
+			timeoutFast,
+			retry,
+		).Should(
+			BeTrue(),
+			"instance in some namespace never got a finalizer",
+		)
+		// And there should also be a namespace annotation.
+		tc.G.Eventually(
+			gotNamespaceAnnotation,
+			timeoutFast,
+			retry,
+		).Should(
+			BeTrue(),
+			"instance in some namespace never got an operator namespace annotation",
+		)
+	} else {
+		t.Log("**** restricted namespaces mode")
+		// We can tell that the resource isn't being reconciled if it
+		// never gets a finalizer.
+		tc.G.Consistently(
```

Comment:
So there isn't a friendly `tc.Consistently` that does the right thing with regard to record/replay and polling interval, but maybe you could write one and then use that? It can't wait for `RemainingTime()` as I don't think that'll work in the CI pass (`RemainingTime() is quite long then, I think?) - but it should be able to use `tc.PollingInterval()` and possibly you can make a nice `Consistently` rule

Reply:
Well, `Consistently` when it's successful ends up waiting for the the full time, so I don't think that can use `RemainingTime`. These aren't ever waiting for ARM, only ever to make sure that the object in question doesn't get a finalizer, so I might just pull it into a function rather than sticking it onto `KubePerTestContext`. Although one downside of that is the no call stack problem... Might chat with you about this one tomorrow.

Reply:
I ended up doing this - I think it's good.